### PR TITLE
refactor + timeout + retry on connectionError

### PR DESCRIPTION
**Pull request by Ragge-dev**
Python part of https://annotell.atlassian.net/browse/INT-748

I did two things in order to try and improve the robustness of the upload:
1. I added a timeout so that the upload/download cannot hang indefinetely
2. I added so that you retry uploads/downloads on a ConnectionError

Note that I did not retry on a Timeout, I felt unsure if this would just make things worse or not.

I also refactored a bit so that upload and download are handled by separate classes, makes for a better structure I think.

### CODE REVIEW

File: annotell-input-api/annotell/input_api/util.py [hunk 1]
```diff
@@ -50,15 +53,29 @@ def get_image_dimensions(image_path: str) -> dict:
...
+    """
+    Calculates the wait time before attempting another file upload or download
+
+    :param upload_attempt: How many attempts to upload that have been made
+    :return: int: The time to wait before retrying upload
+    """
+    max_wait_time = pow(2, upload_attempt - 1)
+    wait_time = random.random() * max_wait_time
+    wait_time = wait_time if wait_time < max_retry_wait_time else max_retry_wait_time
+    return wait_time
```
Comment (alydh-annotell) [inline annotell-input-api/annotell/input_api/util.py:81]: Why do we need to randomize the wait time?

File: annotell-input-api/annotell/input_api/cloud_storage/download_handler.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        :return: JSON deserialized to dictionary
+        """
+        resp = requests.get(url, timeout=self.timeout)
+        try:
+            resp.raise_for_status()
+        except (HTTPError, ConnectionError) as e:
+            http_condition = number_of_retries > 0 and resp.status_code in RETRYABLE_STATUS_CODES
+            if http_condition or isinstance(e, ConnectionError):
+                upload_attempt = self.max_num_retries - number_of_retries + 1
+                wait_time = get_wait_time(upload_attempt, self.max_retry_wait_time)
```
Comment (alydh-annotell) [inline annotell-input-api/annotell/input_api/cloud_storage/download_handler.py:47]:
```suggestion
                download_attempt = self.max_num_retries - number_of_retries + 1
                wait_time = get_wait_time(download_attempt, self.max_retry_wait_time)
```

File: annotell-input-api/annotell/input_api/cloud_storage/upload_handler.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    def _handle_upload_error(self, upload_url: str, file: BinaryIO, headers: Dict[str, str], resp: Response, number_of_retries: int):
+        upload_attempt = self.max_num_retries - number_of_retries + 1
+        log.error(
+            f"On upload attempt ({upload_attempt}/{self.max_num_retries}) to GCS "
+            f"got response:\n{resp.status_code}: {resp.content}"
+        )
+        wait_time = get_wait_time(upload_attempt, self.max_retry_wait_time)
+        log.info(f"Waiting {int(wait_time)} seconds before retrying")
+        time.sleep(wait_time)
+        self._upload_file(upload_url, file, headers, number_of_retries - 1)
```
Comment (alydh-annotell) [inline annotell-input-api/annotell/input_api/cloud_storage/upload_handler.py:55]:
```suggestion
        except (HTTPError, ConnectionError) as e:
            http_condition = number_of_retries > 0 and resp.status_code in RETRYABLE_STATUS_CODES
            if http_condition or isinstance(e, ConnectionError):
                self._handle_upload_error(resp, number_of_retries)
                self._upload_file(upload_url, file, headers, number_of_retries - 1)
            else:
                raise e

    def _handle_upload_error(self, resp: Response, number_of_retries: int):
        upload_attempt = self.max_num_retries - number_of_retries + 1
        log.error(
            f"On upload attempt ({upload_attempt}/{self.max_num_retries}) to GCS "
            f"got response:\n{resp.status_code}: {resp.content}"
        )
        wait_time = get_wait_time(upload_attempt, self.max_retry_wait_time)
        log.info(f"Waiting {int(wait_time)} seconds before retrying")
        time.sleep(wait_time)

```
I think this is neater. That way we don't have to pass all the parameters for uploading to the "waiting" method

Comment (alydh-annotell) [reply annotell-input-api/annotell/input_api/cloud_storage/upload_handler.py:55]: Also removed the unnecessary except handling

**Review by alydh-annotell — COMMENTED** [review]
I like the seperation of the file handling!
Consider my suggestions and then merge :)

File: annotell-input-api/annotell/input_api/util.py (hunk 1 above)
Comment (Ragge-dev) [reply annotell-input-api/annotell/input_api/util.py:81]: Will link to this https://cloud.google.com/iot/docs/how-tos/exponential-backoff

And probably change the implementation a bit to be a bit mor simiilar to the link

File: annotell-input-api/annotell/input_api/cloud_storage/upload_handler.py (hunk 3 above)
Comment (Ragge-dev) [reply annotell-input-api/annotell/input_api/cloud_storage/upload_handler.py:55]: I don't dare to commit this whole suggestion, but I'll be sure to implement it! ;D

**Pull request merged**